File: packages/guard/src/managed.ts
/**
 * Managed Guard Client — OPA/LaunchDarkly-style rule management.
 *
 * Fetches rule bundles from Savants cloud on startup.
 * Evaluates guards locally (zero network latency per check).
 * Reports events in batches (async, non-blocking).
 * Polls for rule changes (ETag-based, 304 on no change).
 */

import type { ASTNode } from './types.js';

// ============================================================
// TYPES
// ============================================================

export interface ManagedOptions {
  apiKey: string;
  apiUrl?: string;
  pollInterval?: number;   // ms, default: 30000
  batchSize?: number;      // default: 50
  batchInterval?: number;  // ms, default: 30000
}

export interface BundleRule {
  id: string;
  dsl: string;
  ast_json: string;
  priority: number;
}

export interface RuleBundle {
  version: number;
  hash: string;
  rules: BundleRule[];
  updated_at: string;
}

export interface ParsedManagedRule {
  id: string;
  dsl: string;
  condition: ASTNode;
  action: string;
  priority: number;
}

export interface GuardEvent {
  context_hash: string;
  action?: string;
  tool?: string;
  result: string;
  matched_rule?: string;
  bundle_version?: number;
  timestamp: string;
}

// ============================================================
// CLIENT
// ============================================================

export class ManagedGuardClient {
  private apiUrl: string;
  private apiKey: string;
  private pollInterval: number;
  private batchSize: number;
  private batchInterval: number;

  private currentHash: string = "";
  private bundleVersion: number = 0;
  private pollTimer: ReturnType<typeof setInterval> | null = null;
  private flushTimer: ReturnType<typeof setInterval> | null = null;
  private eventQueue: GuardEvent[] = [];

  constructor(options: ManagedOptions) {
    this.apiUrl = options.apiUrl || "https://api.savants.cloud";
    this.apiKey = options.apiKey;
    this.pollInterval = options.pollInterval || 30000;
    this.batchSize = options.batchSize || 50;
    this.batchInterval = options.batchInterval || 30000;
  }

  /**
   * Fetch the rule bundle from the cloud. Returns parsed rules.
   */
  async fetchBundle(): Promise<ParsedManagedRule[]> {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
    };
    if (this.currentHash) {
      headers["If-None-Match"] = this.currentHash;
    }

    const resp = await fetch(`${this.apiUrl}/api/v1/guard/bundle`, { headers });

    if (resp.status === 304) {
      return []; // No change, caller keeps existing rules
    }

    if (!resp.ok) {
      throw new Error(`Failed to fetch bundle: ${resp.status}`);
    }

    const bundle = (await resp.json()) as RuleBundle;
    this.currentHash = bundle.hash;
    this.bundleVersion = bundle.version;

    return bundle.rules.map((r) => {
      const parsed = JSON.parse(r.ast_json) as { condition: ASTNode; action: string };
      return {
        id: r.id,
        dsl: r.dsl,
        condition: parsed.condition,
        action: parsed.action,
        priority: r.priority,
      };
    });
  }

  /**
   * Start polling for rule changes.
   * Calls onUpdate when new rules are available.
   */
  startPolling(onUpdate: (rules: ParsedManagedRule[]) => void): void {
    if (this.pollTimer) return;
    this.pollTimer = setInterval(async () => {
      try {
        const rules = await this.fetchBundle();
        if (rules.length > 0) {
          onUpdate(rules);
        }
      } catch {
        // Silent fail on poll — keep using cached rules
      }
    }, this.pollInterval);
  }

  /**
   * Stop polling.
   */
  stopPolling(): void {
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
  }

  /**
   * Queue a guard evaluation event.
   * Flushes automatically when batch is full.
   */
  reportEvent(event: Omit<GuardEvent, "bundle_version">): void {
    this.eventQueue.push({
      ...event,
      bundle_version: this.bundleVersion,
    });

    if (this.eventQueue.length >= this.batchSize) {
      this.flush().catch(() => {}); // Fire and forget
    }
  }

  /**
   * Start the periodic flush timer.
   */
  startFlushing(): void {
    if (this.flushTimer) return;
    this.flushTimer = setInterval(() => {
      if (this.eventQueue.length > 0) {
        this.flush().catch(() => {});
      }
    }, this.batchInterval);
  }

  /**
   * Flush all queued events to the cloud.
   */
  async flush(): Promise<void> {
    if (this.eventQueue.length === 0) return;

    const events = this.eventQueue.splice(0, this.eventQueue.length);

    try {
      await fetch(`${this.apiUrl}/api/v1/guard/events`, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ events }),
      });
    } catch {
      // On failure, put events back in queue for next flush
      this.eventQueue.unshift(...events);
    }
  }

  /**
   * Stop polling + flush remaining events.
   */
  async close(): Promise<void> {
    this.stopPolling();
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
    await this.flush();
  }
}

File: packages/guard-python/savants_guard/managed.py
"""
Managed Guard Client — syncs rules from Savants cloud.

Fetches rule bundles, polls for changes, batches events.
Uses httpx for async HTTP. Only imported when managed=True.
"""

from __future__ import annotations

import asyncio
import json
from datetime import datetime, timezone
from typing import Any, Callable

from .types import ASTNode, ParsedRule


class ManagedGuardClient:
    """Async client for Savants cloud rule management."""

    def __init__(
        self,
        api_key: str,
        api_url: str = "https://api.savants.cloud",
        poll_interval: int = 30,
        batch_size: int = 50,
        batch_interval: int = 30,
    ):
        self._api_key = api_key
        self._api_url = api_url
        self._poll_interval = poll_interval
        self._batch_size = batch_size
        self._batch_interval = batch_interval
        self._current_hash = ""
        self._bundle_version = 0
        self._event_queue: list[dict[str, Any]] = []
        self._poll_task: asyncio.Task[None] | None = None
        self._flush_task: asyncio.Task[None] | None = None

    async def fetch_bundle(self) -> list[ParsedRule]:
        """Fetch rule bundle from cloud. Returns parsed rules or empty list if unchanged."""
        import httpx

        headers = {"Authorization": f"Bearer {self._api_key}"}
        if self._current_hash:
            headers["If-None-Match"] = self._current_hash

        async with httpx.AsyncClient() as client:
            resp = await client.get(
                f"{self._api_url}/api/v1/guard/bundle",
                headers=headers,
                timeout=10,
            )

        if resp.status_code == 304:
            return []

        if resp.status_code != 200:
            raise RuntimeError(f"Failed to fetch bundle: {resp.status_code}")

        data = resp.json()
        self._current_hash = data.get("hash", "")
        self._bundle_version = data.get("version", 0)

        rules: list[ParsedRule] = []
        for r in data.get("rules", []):
            parsed = json.loads(r["ast_json"])
            rules.append(ParsedRule(
                dsl=r["dsl"],
                condition=parsed["condition"],
                action=parsed["action"],
            ))
        return rules

    def start_polling(self, on_update: Callable[[list[ParsedRule]], None]) -> None:
        """Start background polling for rule changes."""
        if self._poll_task is not None:
            return

        async def _poll_loop() -> None:
            while True:
                await asyncio.sleep(self._poll_interval)
                try:
                    rules = await self.fetch_bundle()
                    if rules:
                        on_update(rules)
                except Exception:
                    pass  # Silent fail, keep using cached rules

        self._poll_task = asyncio.get_event_loop().create_task(_poll_loop())

    def stop_polling(self) -> None:
        """Stop polling."""
        if self._poll_task:
            self._poll_task.cancel()
            self._poll_task = None

    def report_event(self, event: dict[str, Any]) -> None:
        """Queue a guard evaluation event. Auto-flushes when batch is full."""
        event["bundle_version"] = self._bundle_version
        self._event_queue.append(event)
        if len(self._event_queue) >= self._batch_size:
            asyncio.get_event_loop().create_task(self.flush())

    def start_flushing(self) -> None:
        """Start periodic flush timer."""
        if self._flush_task is not None:
            return

        async def _flush_loop() -> None:
            while True:
                await asyncio.sleep(self._batch_interval)
                if self._event_queue:
                    await self.flush()

        self._flush_task = asyncio.get_event_loop().create_task(_flush_loop())

    async def flush(self) -> None:
        """Flush queued events to cloud."""
        if not self._event_queue:
            return

        events = self._event_queue[:]
        self._event_queue.clear()

        try:
            import httpx
            async with httpx.AsyncClient() as client:
                await client.post(
                    f"{self._api_url}/api/v1/guard/events",
                    json={"events": events},
                    headers={"Authorization": f"Bearer {self._api_key}",
                             "Content-Type": "application/json"},
                    timeout=5,
                )
        except Exception:
            # Re-queue on failure
            self._event_queue = events + self._event_queue

    async def close(self) -> None:
        """Stop polling and flush remaining events."""
        self.stop_polling()
        if self._flush_task:
            self._flush_task.cancel()
            self._flush_task = None
        await self.flush()
